stan/math/prim/mat/vectorize/apply_scalar_binary.hpp

```diff
@@ -0,0 +1,239 @@
+#ifndef STAN_MATH_PRIM_MAT_VECTORIZE_APPLY_BINARY_SCALAR_HPP
+#define STAN_MATH_PRIM_MAT_VECTORIZE_APPLY_BINARY_SCALAR_HPP
+
+#include <stan/math/prim/mat/err/check_matching_dims.hpp>
+#include <stan/math/prim/scal/err/check_size_match.hpp>
+#include <Eigen/Dense>
+#include <vector>
+
+namespace stan {
+
+  namespace math {
+
+    /**
+     * Base template class for vectorization of binary scalar functions
+     * defined by a template class <code>F</code> to a scalar,
+     * standard library vector, or Eigen dense matrix expression
+     * template.
+     *
+     * <p>The base class applies to any Eigen dense matrix expression
+     * template.  Specializations define applications to scalars
+     * (primitive or autodiff in the corresponding autodiff library
+     * directories) or to standard library vectors of vectorizable
+     * types (primitives, Eigen dense matrix expressions, or further
+     * standard vectors).
+     *
+     * <p>Each specialization must define the typedef
+     * <code>return_t</code> for the vectorized return type and the
+     * function <code>apply</code> which defines the vectorization or
+     * base application of the function defined statically by the
+     * class F.  The function definition class F defines a static
+     * function <code>fun()</code>, which defines the function's
+     * behavior on scalars.
+     *
+     * @tparam F Type of function to apply.
+     * @tparam T1 Type of first argument to which function is applied.
+     * @tparam T2 Type of second argument to which function is applied.
+     */
+    template <typename F, typename T1, typename T2>
+    struct apply_scalar_binary {
+      /**
+       * Type of underlying scalar for the matrix type T1.
+       * Type of underlying scalar for the matrix type T2.
+       */
+      typedef typename Eigen::internal::traits<T1>::Scalar scalar_t1;
+      typedef typename Eigen::internal::traits<T2>::Scalar scalar_t2;
+
+      /**
+       * Return type for applying the function elementwise to a matrix
+       * expression template of type T1 and T2.
+       */
+      typedef Eigen::Matrix<typename apply_scalar_binary<F, scalar_t1,
+      scalar_t2>::return_t, T1::RowsAtCompileTime, T1::ColsAtCompileTime>
+      return_t;
+
+      /**
+       * Return the result of applying the function defined by the
+       * template parameter F to the specified matrix argument.
+       *
+       * @param x Matrix to which operation is applied.
+       * @param y Matrix to which operation is applied.
+       * @return Componentwise application of the function specified
+       * by F to the specified matrix.
+       */
+      static inline return_t apply(const T1& x, const T2& y) {
+        check_matching_dims<scalar_t1, scalar_t2,
```

Comment:
Same here.  You should read the Google code style guide and try to follow it as much as possible!